lib/WUI/nhttp/send_json.cpp

```diff
@@ -0,0 +1,128 @@
+#include "send_json.h"
+#include "handler.h"
+#include "headers.h"
+
+#include <timing.h>
+#include <segmented_json_macros.h>
+#include <http/chunked.h>
+#include <transfers/monitor.hpp>
+
+using transfers::Monitor;
+using namespace json;
+using namespace http;
+
+namespace nhttp::handler {
+
+json::JsonResult EmptyRenderer::renderState(size_t resume_point, json::JsonOutput &output, Empty &state) const {
+    return generator(resume_point, output);
+}
+
+JsonResult TransferRenderer::renderState(size_t resume_point, json::JsonOutput &output, TransferState &state) const {
+    // Note: We are guaranteed to get a status, because we checked if it is present
+    // before calling this and we set allow_stale to true
+    auto transfer_status = Monitor::instance.status(true);
+    // Keep the indentation of the JSON in here!
+    // clang-format off
+    JSON_START;
+        JSON_OBJ_START
+            if (transfer_status->id != state.transfer_id) {
+                // if transfer changes mid report, bail out
+                transfer_status.reset();
+            }
+
+            JSON_FIELD_STR("target", "usb") JSON_COMMA;
+            JSON_FIELD_STR_FORMAT_G(transfer_status.has_value(), "type", "%s", to_str(transfer_status->type)) JSON_COMMA;
+            JSON_FIELD_STR_FORMAT_G(transfer_status.has_value(), "size", "%d", transfer_status->expected) JSON_COMMA;
+            //FIXME: Right now should be seconds from epoch start, would be much nicer
+            // for us to make it seconds from print start and make the client compute the rest,
+            // also it would solve some potential problems with NTP and time zones etc.
+            JSON_FIELD_STR_FORMAT_G(transfer_status.has_value(), "start_time", "%d", time(nullptr) - (ticks_s() - transfer_status->start)) JSON_COMMA;
+            JSON_FIELD_BOOL_G(transfer_status.has_value(), "to_print", transfer_status->print_after_upload) JSON_COMMA;
+            // Note: This works, because destination cannot go from non null to null
+            // (if one transfer ends and another starts mid report, we bail out)
+            if (transfer_status->destination) {
+                JSON_FIELD_STR_G(transfer_status.has_value(), "destination", transfer_status->destination) JSON_COMMA;
+            }
+            // To prevent division by zero
+            if (transfer_status->transferred !=0 && ticks_s() > transfer_status->start) {
```

Comment:
I'm not sure this is 100% correct, for several reasons:
* The field may be retried with a different value.
* The time may (and does) overflow (or doesn't it, in case of seconds?)

I'd actually go with something like `... / std::min(1, transfer_status->expected)`

Reply:
`ticks_s` overflows every 136 years, so we are fine there, so the only time we are in danger of division by zero is in the very beginning, where no data was transferred or the very first second of the transfer. 

IMHO this is secure, because once those checks passed once, they can never stop being true. Neither transferred nor ticks_s can ever decrease, nevertheless your solution may be more explicitly correct.